File: fvm/src-f/poi_gen.f
!C
!C***
!C*** POI_GEN
!C***
!C
!C    generate COEF. MATRIX for POISSON equations
!C    
      subroutine POI_GEN

      use STRUCT
      use PCG

      implicit REAL*8 (A-H,O-Z)

!C
!C-- INIT.
      nn = ICELTOT

      NLU= 6

      allocate (BFORCE(nn), D(nn), PHI(nn))
      allocate (INLU(nn))

      PHI   = 0.d0
      BFORCE= 0.d0
        D   = 0.d0

      INLU= 0

!C
!C-- INTERIOR & NEUMANN boundary's
      do icel= 1, ICELTOT
        icN1= NEIBcell(icel,1)
        icN2= NEIBcell(icel,2)
        icN3= NEIBcell(icel,3)
        icN4= NEIBcell(icel,4)
        icN5= NEIBcell(icel,5)
        icN6= NEIBcell(icel,6)

        if (icN5.ne.0) then
          INLU(icel)= INLU(icel) + 1
        endif

        if (icN3.ne.0) then
          INLU(icel)= INLU(icel) + 1           
        endif

        if (icN1.ne.0) then
          INLU(icel)= INLU(icel) + 1           
        endif

        if (icN2.ne.0) then
          INLU(icel)= INLU(icel) + 1           
        endif

        if (icN4.ne.0) then
          INLU(icel)= INLU(icel) + 1           
        endif

        if (icN6.ne.0) then
          INLU(icel)= INLU(icel) + 1           
        endif

      enddo

!C
!C-- 1D array
      allocate (indexLU(0:nn))
      indexLU= 0
     
      do icel= 1, ICELTOT
        indexLU(icel)= INLU(icel)
      enddo

      do icel= 1, ICELTOT
        indexLU(icel)= indexLU(icel) + indexLU(icel-1)
      enddo

      NPLU= indexLU(ICELTOT)

      allocate (itemLU(NPLU), AMAT(NPLU))

      itemLU= 0
      AMAT= 0.d0
!C===

!C
!C +-----------------------------------+
!C | INTERIOR & NEUMANN BOUNDARY CELLs |
!C +-----------------------------------+
!C===
      icouG= 0
      do icel= 1, ICELTOT
        icN1= NEIBcell(icel,1)
        icN2= NEIBcell(icel,2)
        icN3= NEIBcell(icel,3)
        icN4= NEIBcell(icel,4)
        icN5= NEIBcell(icel,5)
        icN6= NEIBcell(icel,6)


        VOL0= VOLCEL(icel)

        icou= 0
        if (icN5.ne.0) then
          coef   =RDZ * ZAREA
          D(icel)= D(icel) - coef
                   icou= icou + 1
                   k   = icou + indexLU(icel-1)
          itemLU(k)= icN5
            AMAT(k)= coef
        endif

        if (icN3.ne.0) then
          coef   = RDY * YAREA
          D(icel)= D(icel) - coef
                   icou= icou + 1
                   k   = icou + indexLU(icel-1)
          itemLU(k)= icN3
            AMAT(k)= coef
        endif

        if (icN1.ne.0) then
          coef   = RDX * XAREA
          D(icel)= D(icel) - coef
                   icou= icou + 1        
                   k   = icou + indexLU(icel-1)
          itemLU(k)= icN1
            AMAT(k)= coef
        endif

        if (icN2.ne.0) then
          coef   = RDX * XAREA
          D(icel)= D(icel) - coef
                   icou= icou + 1
                   k   = icou + indexLU(icel-1)
          itemLU(k)= icN2
            AMAT(k)= coef
        endif

        if (icN4.ne.0) then
          coef   = RDY * YAREA
          D(icel)= D(icel) - coef
                   icou= icou + 1
                   k   = icou + indexLU(icel-1)
          itemLU(k)= icN4
            AMAT(k)= coef
        endif

        if (icN6.ne.0) then
          coef   = RDZ * ZAREA
          D(icel)= D(icel) - coef
                   icou= icou + 1
                   k   = icou + indexLU(icel-1)
          itemLU(k)= icN6
            AMAT(k)= coef
        endif

        ii= XYZ(icel,1)
        jj= XYZ(icel,2)
        kk= XYZ(icel,3)

        BFORCE(icel)= -dfloat(ii+jj+kk) * VOL0

      enddo
!C===

!C
!C +--------------------------+
!C | DIRICHLET BOUNDARY CELLs |
!C +--------------------------+
!C   TOP SURFACE
!C===
      do ib= 1, ZmaxCELtot
        icel= ZmaxCEL(ib)
        coef= 2.d0 * RDZ * ZAREA
        D(icel)= D(icel) - coef
      enddo
!C===

      return
      end

File: fvm/src-c/poi_gen.c
/*
 * POI_GEN
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>
#include <math.h>

#include "struct_ext.h"
#include "pcg_ext.h"
#include "poi_gen.h"
#include "allocate.h"

extern int
POI_GEN(void)
{
	int nn;
	int ic0, icN1, icN2, icN3, icN4, icN5, icN6;
	int i, j, k, ib, ic, ip, icel, icou, icol, icouG;
	int ii, jj, kk, nn1, num, nr, j0, j1;
	double coef, VOL0, S1t, E1t;
        int isLU;


	NLU= 6;

	BFORCE = (double *)allocate_vector(sizeof(double),ICELTOT);
	D      = (double *)allocate_vector(sizeof(double),ICELTOT);
	PHI    = (double *)allocate_vector(sizeof(double),ICELTOT);
        INLU = (int *)allocate_vector(sizeof(int),ICELTOT);

        indexLU= (int *)allocate_vector(sizeof(int),ICELTOT+1);

        for (i = 0; i <ICELTOT ; i++) {
		BFORCE[i]=0.0;
		D[i]  =0.0;
		PHI[i]=0.0;
        	INLU[i] = 0;
        }

        for (i = 0; i <=ICELTOT ; i++) {
		indexLU[i] = 0;
	}


/*********************************
 * INTERIOR & NEUMANN boundary's *
 *********************************/

	for(icel=0; icel<ICELTOT; icel++) {
		icN1 = NEIBcell[icel][0];
		icN2 = NEIBcell[icel][1];
		icN3 = NEIBcell[icel][2];
		icN4 = NEIBcell[icel][3];
		icN5 = NEIBcell[icel][4];
		icN6 = NEIBcell[icel][5];

		if(icN5 != 0) {
 		   INLU[icel]= INLU[icel] + 1;		  
		}

		if(icN3 != 0) {
 		   INLU[icel]= INLU[icel] + 1;
		}

		if(icN1 != 0) {
 		   INLU[icel]= INLU[icel] + 1;		  
		}

		if(icN2 != 0) {
 		   INLU[icel]= INLU[icel] + 1;		  
		}

		if(icN4 != 0) {
 		   INLU[icel]= INLU[icel] + 1;		  
		}

		if(icN6 != 0) {
 		   INLU[icel]= INLU[icel] + 1;		  
		}
	}


/********************************************
* 1D ordering: indexL, indexU, itemL, itemU *
*********************************************/

        for(i=0; i<ICELTOT; i++){
                indexLU[i+1]=indexLU[i]+INLU[i];
        }
        NPLU= indexLU[ICELTOT];

        itemLU= (int *)allocate_vector(sizeof(int),NPLU);
        AMAT  = (double *)allocate_vector(sizeof(double),NPLU);

	for(i=0; i<ICELTOT; i++) {
	  for(j=indexLU[i]; j<indexLU[i+1]; j++) {
	    itemLU[j]=0;
	      AMAT[j]=0.0;
	  }
	}

        free(INLU);

/*************************************
 * INTERIOR & NEUMANN BOUNDARY CELLs *
 *************************************/
	for(icel=0; icel<ICELTOT; icel++) {
			icN1 = NEIBcell[icel][0];
			icN2 = NEIBcell[icel][1];
			icN3 = NEIBcell[icel][2];
			icN4 = NEIBcell[icel][3];
			icN5 = NEIBcell[icel][4];
			icN6 = NEIBcell[icel][5];

	                VOL0 = VOLCEL[icel];

       		        isLU=indexLU[icel];

                        icou= 0;
			if(icN5 != 0) {
				coef = RDZ * ZAREA;
				D[icel] -= coef;
				itemLU[icou+isLU]= icN5-1;
				AMAT  [icou+isLU]= coef;
				icou= icou + 1;
			}

			if(icN3 != 0) {
				coef = RDZ * YAREA;
				D[icel] -= coef;
				itemLU[icou+isLU]= icN3-1;
				AMAT  [icou+isLU]= coef;
				icou= icou + 1;
			}

			if(icN1 != 0) {
				coef = RDZ * XAREA;
				D[icel] -= coef;
				itemLU[icou+isLU]= icN1-1;
				AMAT  [icou+isLU]= coef;
				icou= icou + 1;
			}

			if(icN2 != 0) {
				coef = RDZ * XAREA;
				D[icel] -= coef;
				itemLU[icou+isLU]= icN2-1;
				AMAT  [icou+isLU]= coef;
				icou= icou + 1;
			}

			if(icN4 != 0) {
				coef = RDZ * YAREA;
				D[icel] -= coef;
				itemLU[icou+isLU]= icN4-1;
				AMAT  [icou+isLU]= coef;
				icou= icou + 1;
			}

			if(icN6 != 0) {
			        coef = RDZ * ZAREA;
				D[icel] -= coef;
				itemLU[icou+isLU]= icN6-1;
				AMAT  [icou+isLU]= coef;
				icou= icou + 1;
			}

			ii = XYZ[icel][0];
			jj = XYZ[icel][1];
			kk = XYZ[icel][2];

			BFORCE[icel] = - (double)(ii + jj + kk) * VOLCEL[icel];
		}

/****************************
 * DIRICHLET BOUNDARY CELLs *
 ****************************/
/* TOP SURFACE */
	for(ib=0; ib<ZmaxCELtot; ib++) {
		icel  = ZmaxCEL[ib];
		coef = 2.0 * RDZ * ZAREA;
		D[icel-1] -= coef;
	}


	return 0;
}
